{
  "nbformat": 4,
  "nbformat_minor": 0,
  "metadata": {
    "colab": {
      "name": "GSE135922_GSM4037981.ipynb",
      "provenance": [],
      "collapsed_sections": [],
      "include_colab_link": true
    },
    "kernelspec": {
      "display_name": "Python 3",
      "name": "python3"
    }
  },
  "cells": [
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "view-in-github",
        "colab_type": "text"
      },
      "source": [
        "<a href=\"https://colab.research.google.com/github/agalvezm/ACE2_scRNAseq/blob/master/tcc/GSE135922_GSM4037981_from_gc.ipynb\" target=\"_parent\"><img src=\"https://colab.research.google.com/assets/colab-badge.svg\" alt=\"Open In Colab\"/></a>"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "V9ryXluJsJw9"
      },
      "source": [
        "# GSE135922_GSM4037981"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "g9-EHujcVc8a"
      },
      "source": [
        "This notebook uses the filtered count matrix resulting from the following google colab notebook. https://github.com/agalvezm/ACE2_scRNAseq/blob/master/notebooks_countmatrices/GSE135922_GSM4037981.ipynb Please run the notebook above and upload the results file in /content"
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "2BiFt2wu4aLh",
        "outputId": "f863e40c-5b87-4ec0-a5d4-0b83956d1bdb"
      },
      "source": [
        "# Install SRA-toolkit \n",
        "!wget \"http://ftp-trace.ncbi.nlm.nih.gov/sra/sdk/current/sratoolkit.current-centos_linux64.tar.gz\"\n",
        "\n",
        "! tar -xzf sratoolkit.current-centos_linux64.tar.gz\n",
        "\n",
        "# Add to path\n",
        "import os\n",
        "os.environ['PATH'] += \":/content/sratoolkit.2.11.0-centos_linux64/bin\"\n",
        "\n",
        "# Configure\n",
        "!vdb-config --interactive\n",
        "\n",
        "# Import packages\n",
        "\n",
        "import numpy as np\n"
      ],
      "execution_count": 1,
      "outputs": [
        {
          "output_type": "stream",
          "text": [
            "--2021-04-12 17:48:45--  http://ftp-trace.ncbi.nlm.nih.gov/sra/sdk/current/sratoolkit.current-centos_linux64.tar.gz\n",
            "Resolving ftp-trace.ncbi.nlm.nih.gov (ftp-trace.ncbi.nlm.nih.gov)... 130.14.250.13, 130.14.250.11, 2607:f220:41e:250::11, ...\n",
            "Connecting to ftp-trace.ncbi.nlm.nih.gov (ftp-trace.ncbi.nlm.nih.gov)|130.14.250.13|:80... connected.\n",
            "HTTP request sent, awaiting response... 301 Moved Permanently\n",
            "Location: https://ftp-trace.ncbi.nlm.nih.gov/sra/sdk/current/sratoolkit.current-centos_linux64.tar.gz [following]\n",
            "--2021-04-12 17:48:46--  https://ftp-trace.ncbi.nlm.nih.gov/sra/sdk/current/sratoolkit.current-centos_linux64.tar.gz\n",
            "Connecting to ftp-trace.ncbi.nlm.nih.gov (ftp-trace.ncbi.nlm.nih.gov)|130.14.250.13|:443... connected.\n",
            "HTTP request sent, awaiting response... 200 OK\n",
            "Length: 99139357 (95M) [application/x-gzip]\n",
            "Saving to: ‘sratoolkit.current-centos_linux64.tar.gz’\n",
            "\n",
            "sratoolkit.current- 100%[===================>]  94.55M   354MB/s    in 0.3s    \n",
            "\n",
            "2021-04-12 17:48:46 (354 MB/s) - ‘sratoolkit.current-centos_linux64.tar.gz’ saved [99139357/99139357]\n",
            "\n",
            "\u001b[2J\u001b[?25l\u001b[?1000h\u001b[?1002h2021-04-12T17:48:50 vdb-config.2.11.0 fatal: SIGNAL - Segmentation fault \n"
          ],
          "name": "stdout"
        }
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "mU4XegJlo2Na"
      },
      "source": [
        "# List of SRAs to download and transform. Copy paste from excel as string\n",
        "SRAs = \"SRR9990627\tSRR9990628\tSRR9990629\tSRR9990630\tSRR9990631\tSRR9990632\tSRR9990633\tSRR9990634\tSRR9990635\tSRR9990636\tSRR9990637\tSRR9990638\tSRR9990639\tSRR9990640\tSRR9990641\tSRR9990642\tSRR9990643\tSRR9990644\tSRR9990645\tSRR9990646\tSRR9990647\tSRR9990648\tSRR9990649\tSRR9990650\"\n",
        "\n",
        "SRAs = SRAs.split()\n",
        "\n",
        "# SRA to delete, usually the index file, which get downloaded because we include technical reads. \n",
        "# It varies so check in the SRA website\n",
        "SRA_to_delete = \"3\"\n",
        "\n",
        "\n",
        "SRAs_to_keep = np.setdiff1d([\"1\", \"2\", \"3\"], [SRA_to_delete])\n",
        "\n",
        "# sample specific metadata\n",
        "\n",
        "samp_id = [\"GSM4037981\"]\n",
        "\n",
        "cell_type = [\"retinal pigment epithelium macula, unenriched\"]\n",
        "\n",
        "condition = [\"donor 1\"]"
      ],
      "execution_count": 2,
      "outputs": []
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "u3uDp_jCpfhZ",
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "outputId": "968c77e5-f990-4615-d855-602319d5ae42"
      },
      "source": [
        "for sra in SRAs:\n",
        "  \"\"\"\n",
        "  Downloads SRAs and transforms to compressed fastq file. Remove index reads and original SRA\n",
        "  \"\"\"\n",
        "  # Download SRA\n",
        "  !prefetch $sra\n",
        "\n",
        "  # Convert to fastq. Biological reads are usually considered technicals in 10x so do not skip them.\n",
        "  fastq_dump_cmd = \"fastq-dump -F --gzip --readids --split-files \" + sra + \"/\" + sra +\".sra\"\n",
        "  !$fastq_dump_cmd\n",
        "\n",
        "  # Remove files\n",
        "  rm_cmd_1 = \"rm \" + sra + \"/\" + sra + \".sra\"\n",
        "  rm_cmd_2 = \"rm \" + sra + \"_\" + SRA_to_delete + \".fastq.gz\" \n",
        "  !$rm_cmd_1\n",
        "  !$rm_cmd_2"
      ],
      "execution_count": 3,
      "outputs": [
        {
          "output_type": "stream",
          "text": [
            "\n",
            "2021-04-12T17:49:54 prefetch.2.11.0: 1) Downloading 'SRR9990627'...\n",
            "2021-04-12T17:49:54 prefetch.2.11.0:  Downloading via HTTPS...\n",
            "2021-04-12T17:50:28 prefetch.2.11.0:  HTTPS download succeed\n",
            "2021-04-12T17:50:31 prefetch.2.11.0:  'SRR9990627' is valid\n",
            "2021-04-12T17:50:31 prefetch.2.11.0: 1) 'SRR9990627' was downloaded successfully\n",
            "2021-04-12T17:50:31 prefetch.2.11.0: 'SRR9990627' has 0 unresolved dependencies\n",
            "Read 10536649 spots for SRR9990627/SRR9990627.sra\n",
            "Written 10536649 spots for SRR9990627/SRR9990627.sra\n",
            "rm: cannot remove 'SRR9990627_3.fastq.gz': No such file or directory\n",
            "\n",
            "2021-04-12T18:07:28 prefetch.2.11.0: 1) Downloading 'SRR9990628'...\n",
            "2021-04-12T18:07:28 prefetch.2.11.0:  Downloading via HTTPS...\n",
            "2021-04-12T18:08:04 prefetch.2.11.0:  HTTPS download succeed\n",
            "2021-04-12T18:08:07 prefetch.2.11.0:  'SRR9990628' is valid\n",
            "2021-04-12T18:08:07 prefetch.2.11.0: 1) 'SRR9990628' was downloaded successfully\n",
            "2021-04-12T18:08:07 prefetch.2.11.0: 'SRR9990628' has 0 unresolved dependencies\n",
            "Read 11329458 spots for SRR9990628/SRR9990628.sra\n",
            "Written 11329458 spots for SRR9990628/SRR9990628.sra\n",
            "rm: cannot remove 'SRR9990628_3.fastq.gz': No such file or directory\n",
            "\n",
            "2021-04-12T18:26:12 prefetch.2.11.0: 1) Downloading 'SRR9990629'...\n",
            "2021-04-12T18:26:12 prefetch.2.11.0:  Downloading via HTTPS...\n",
            "2021-04-12T18:26:46 prefetch.2.11.0:  HTTPS download succeed\n",
            "2021-04-12T18:26:49 prefetch.2.11.0:  'SRR9990629' is valid\n",
            "2021-04-12T18:26:49 prefetch.2.11.0: 1) 'SRR9990629' was downloaded successfully\n",
            "2021-04-12T18:26:49 prefetch.2.11.0: 'SRR9990629' has 0 unresolved dependencies\n",
            "Read 10724345 spots for SRR9990629/SRR9990629.sra\n",
            "Written 10724345 spots for SRR9990629/SRR9990629.sra\n",
            "rm: cannot remove 'SRR9990629_3.fastq.gz': No such file or directory\n",
            "\n",
            "2021-04-12T18:43:51 prefetch.2.11.0: 1) Downloading 'SRR9990630'...\n",
            "2021-04-12T18:43:51 prefetch.2.11.0:  Downloading via HTTPS...\n",
            "2021-04-12T18:44:24 prefetch.2.11.0:  HTTPS download succeed\n",
            "2021-04-12T18:44:27 prefetch.2.11.0:  'SRR9990630' is valid\n",
            "2021-04-12T18:44:27 prefetch.2.11.0: 1) 'SRR9990630' was downloaded successfully\n",
            "2021-04-12T18:44:27 prefetch.2.11.0: 'SRR9990630' has 0 unresolved dependencies\n",
            "Read 11004293 spots for SRR9990630/SRR9990630.sra\n",
            "Written 11004293 spots for SRR9990630/SRR9990630.sra\n",
            "rm: cannot remove 'SRR9990630_3.fastq.gz': No such file or directory\n",
            "\n",
            "2021-04-12T19:02:11 prefetch.2.11.0: 1) Downloading 'SRR9990631'...\n",
            "2021-04-12T19:02:11 prefetch.2.11.0:  Downloading via HTTPS...\n",
            "2021-04-12T19:02:43 prefetch.2.11.0:  HTTPS download succeed\n",
            "2021-04-12T19:02:46 prefetch.2.11.0:  'SRR9990631' is valid\n",
            "2021-04-12T19:02:46 prefetch.2.11.0: 1) 'SRR9990631' was downloaded successfully\n",
            "2021-04-12T19:02:46 prefetch.2.11.0: 'SRR9990631' has 0 unresolved dependencies\n",
            "Read 10482615 spots for SRR9990631/SRR9990631.sra\n",
            "Written 10482615 spots for SRR9990631/SRR9990631.sra\n",
            "rm: cannot remove 'SRR9990631_3.fastq.gz': No such file or directory\n",
            "\n",
            "2021-04-12T19:19:40 prefetch.2.11.0: 1) Downloading 'SRR9990632'...\n",
            "2021-04-12T19:19:40 prefetch.2.11.0:  Downloading via HTTPS...\n",
            "2021-04-12T19:20:28 prefetch.2.11.0:  HTTPS download succeed\n",
            "2021-04-12T19:20:32 prefetch.2.11.0:  'SRR9990632' is valid\n",
            "2021-04-12T19:20:32 prefetch.2.11.0: 1) 'SRR9990632' was downloaded successfully\n",
            "2021-04-12T19:20:32 prefetch.2.11.0: 'SRR9990632' has 0 unresolved dependencies\n",
            "Read 13343813 spots for SRR9990632/SRR9990632.sra\n",
            "Written 13343813 spots for SRR9990632/SRR9990632.sra\n",
            "rm: cannot remove 'SRR9990632_3.fastq.gz': No such file or directory\n",
            "\n",
            "2021-04-12T19:41:36 prefetch.2.11.0: 1) Downloading 'SRR9990633'...\n",
            "2021-04-12T19:41:36 prefetch.2.11.0:  Downloading via HTTPS...\n",
            "2021-04-12T19:42:09 prefetch.2.11.0:  HTTPS download succeed\n",
            "2021-04-12T19:42:12 prefetch.2.11.0:  'SRR9990633' is valid\n",
            "2021-04-12T19:42:12 prefetch.2.11.0: 1) 'SRR9990633' was downloaded successfully\n",
            "2021-04-12T19:42:12 prefetch.2.11.0: 'SRR9990633' has 0 unresolved dependencies\n",
            "Read 10573404 spots for SRR9990633/SRR9990633.sra\n",
            "Written 10573404 spots for SRR9990633/SRR9990633.sra\n",
            "rm: cannot remove 'SRR9990633_3.fastq.gz': No such file or directory\n",
            "\n",
            "2021-04-12T19:59:17 prefetch.2.11.0: 1) Downloading 'SRR9990634'...\n",
            "2021-04-12T19:59:17 prefetch.2.11.0:  Downloading via HTTPS...\n",
            "2021-04-12T19:59:51 prefetch.2.11.0:  HTTPS download succeed\n",
            "2021-04-12T19:59:54 prefetch.2.11.0:  'SRR9990634' is valid\n",
            "2021-04-12T19:59:54 prefetch.2.11.0: 1) 'SRR9990634' was downloaded successfully\n",
            "2021-04-12T19:59:54 prefetch.2.11.0: 'SRR9990634' has 0 unresolved dependencies\n",
            "Read 11353324 spots for SRR9990634/SRR9990634.sra\n",
            "Written 11353324 spots for SRR9990634/SRR9990634.sra\n",
            "rm: cannot remove 'SRR9990634_3.fastq.gz': No such file or directory\n",
            "\n",
            "2021-04-12T20:18:04 prefetch.2.11.0: 1) Downloading 'SRR9990635'...\n",
            "2021-04-12T20:18:04 prefetch.2.11.0:  Downloading via HTTPS...\n",
            "2021-04-12T20:18:39 prefetch.2.11.0:  HTTPS download succeed\n",
            "2021-04-12T20:18:42 prefetch.2.11.0:  'SRR9990635' is valid\n",
            "2021-04-12T20:18:42 prefetch.2.11.0: 1) 'SRR9990635' was downloaded successfully\n",
            "2021-04-12T20:18:42 prefetch.2.11.0: 'SRR9990635' has 0 unresolved dependencies\n",
            "Read 10843813 spots for SRR9990635/SRR9990635.sra\n",
            "Written 10843813 spots for SRR9990635/SRR9990635.sra\n",
            "rm: cannot remove 'SRR9990635_3.fastq.gz': No such file or directory\n",
            "\n",
            "2021-04-12T20:36:06 prefetch.2.11.0: 1) Downloading 'SRR9990636'...\n",
            "2021-04-12T20:36:06 prefetch.2.11.0:  Downloading via HTTPS...\n",
            "2021-04-12T20:36:32 prefetch.2.11.0:  HTTPS download succeed\n",
            "2021-04-12T20:36:35 prefetch.2.11.0:  'SRR9990636' is valid\n",
            "2021-04-12T20:36:35 prefetch.2.11.0: 1) 'SRR9990636' was downloaded successfully\n",
            "2021-04-12T20:36:35 prefetch.2.11.0: 'SRR9990636' has 0 unresolved dependencies\n",
            "Read 8781311 spots for SRR9990636/SRR9990636.sra\n",
            "Written 8781311 spots for SRR9990636/SRR9990636.sra\n",
            "rm: cannot remove 'SRR9990636_3.fastq.gz': No such file or directory\n",
            "\n",
            "2021-04-12T20:52:06 prefetch.2.11.0: 1) Downloading 'SRR9990637'...\n",
            "2021-04-12T20:52:06 prefetch.2.11.0:  Downloading via HTTPS...\n",
            "2021-04-12T20:52:30 prefetch.2.11.0:  HTTPS download succeed\n",
            "2021-04-12T20:52:32 prefetch.2.11.0:  'SRR9990637' is valid\n",
            "2021-04-12T20:52:32 prefetch.2.11.0: 1) 'SRR9990637' was downloaded successfully\n",
            "2021-04-12T20:52:32 prefetch.2.11.0: 'SRR9990637' has 0 unresolved dependencies\n",
            "Read 7119346 spots for SRR9990637/SRR9990637.sra\n",
            "Written 7119346 spots for SRR9990637/SRR9990637.sra\n",
            "rm: cannot remove 'SRR9990637_3.fastq.gz': No such file or directory\n",
            "\n",
            "2021-04-12T21:05:22 prefetch.2.11.0: 1) Downloading 'SRR9990638'...\n",
            "2021-04-12T21:05:22 prefetch.2.11.0:  Downloading via HTTPS...\n",
            "2021-04-12T21:05:39 prefetch.2.11.0:  HTTPS download succeed\n",
            "2021-04-12T21:05:41 prefetch.2.11.0:  'SRR9990638' is valid\n",
            "2021-04-12T21:05:41 prefetch.2.11.0: 1) 'SRR9990638' was downloaded successfully\n",
            "2021-04-12T21:05:41 prefetch.2.11.0: 'SRR9990638' has 0 unresolved dependencies\n",
            "Read 13115661 spots for SRR9990638/SRR9990638.sra\n",
            "Written 13115661 spots for SRR9990638/SRR9990638.sra\n",
            "rm: cannot remove 'SRR9990638_3.fastq.gz': No such file or directory\n",
            "\n",
            "2021-04-12T21:19:33 prefetch.2.11.0: 1) Downloading 'SRR9990639'...\n",
            "2021-04-12T21:19:33 prefetch.2.11.0:  Downloading via HTTPS...\n",
            "2021-04-12T21:19:55 prefetch.2.11.0:  HTTPS download succeed\n",
            "2021-04-12T21:19:58 prefetch.2.11.0:  'SRR9990639' is valid\n",
            "2021-04-12T21:19:58 prefetch.2.11.0: 1) 'SRR9990639' was downloaded successfully\n",
            "2021-04-12T21:19:58 prefetch.2.11.0: 'SRR9990639' has 0 unresolved dependencies\n",
            "Read 7623882 spots for SRR9990639/SRR9990639.sra\n",
            "Written 7623882 spots for SRR9990639/SRR9990639.sra\n",
            "rm: cannot remove 'SRR9990639_3.fastq.gz': No such file or directory\n",
            "\n",
            "2021-04-12T21:33:36 prefetch.2.11.0: 1) Downloading 'SRR9990640'...\n",
            "2021-04-12T21:33:36 prefetch.2.11.0:  Downloading via HTTPS...\n",
            "2021-04-12T21:33:57 prefetch.2.11.0:  HTTPS download succeed\n",
            "2021-04-12T21:34:00 prefetch.2.11.0:  'SRR9990640' is valid\n",
            "2021-04-12T21:34:00 prefetch.2.11.0: 1) 'SRR9990640' was downloaded successfully\n",
            "2021-04-12T21:34:00 prefetch.2.11.0: 'SRR9990640' has 0 unresolved dependencies\n",
            "Read 7158711 spots for SRR9990640/SRR9990640.sra\n",
            "Written 7158711 spots for SRR9990640/SRR9990640.sra\n",
            "rm: cannot remove 'SRR9990640_3.fastq.gz': No such file or directory\n",
            "\n",
            "2021-04-12T21:46:56 prefetch.2.11.0: 1) Downloading 'SRR9990641'...\n",
            "2021-04-12T21:46:56 prefetch.2.11.0:  Downloading via HTTPS...\n",
            "2021-04-12T21:47:33 prefetch.2.11.0:  HTTPS download succeed\n",
            "2021-04-12T21:47:37 prefetch.2.11.0:  'SRR9990641' is valid\n",
            "2021-04-12T21:47:37 prefetch.2.11.0: 1) 'SRR9990641' was downloaded successfully\n",
            "2021-04-12T21:47:37 prefetch.2.11.0: 'SRR9990641' has 0 unresolved dependencies\n",
            "Read 12937995 spots for SRR9990641/SRR9990641.sra\n",
            "Written 12937995 spots for SRR9990641/SRR9990641.sra\n",
            "rm: cannot remove 'SRR9990641_3.fastq.gz': No such file or directory\n",
            "\n",
            "2021-04-12T22:08:06 prefetch.2.11.0: 1) Downloading 'SRR9990642'...\n",
            "2021-04-12T22:08:06 prefetch.2.11.0:  Downloading via HTTPS...\n",
            "2021-04-12T22:08:37 prefetch.2.11.0:  HTTPS download succeed\n",
            "2021-04-12T22:08:40 prefetch.2.11.0:  'SRR9990642' is valid\n",
            "2021-04-12T22:08:40 prefetch.2.11.0: 1) 'SRR9990642' was downloaded successfully\n",
            "2021-04-12T22:08:40 prefetch.2.11.0: 'SRR9990642' has 0 unresolved dependencies\n",
            "Read 10231477 spots for SRR9990642/SRR9990642.sra\n",
            "Written 10231477 spots for SRR9990642/SRR9990642.sra\n",
            "rm: cannot remove 'SRR9990642_3.fastq.gz': No such file or directory\n",
            "\n",
            "2021-04-12T22:25:19 prefetch.2.11.0: 1) Downloading 'SRR9990643'...\n",
            "2021-04-12T22:25:19 prefetch.2.11.0:  Downloading via HTTPS...\n",
            "2021-04-12T22:25:53 prefetch.2.11.0:  HTTPS download succeed\n",
            "2021-04-12T22:25:56 prefetch.2.11.0:  'SRR9990643' is valid\n",
            "2021-04-12T22:25:56 prefetch.2.11.0: 1) 'SRR9990643' was downloaded successfully\n",
            "2021-04-12T22:25:56 prefetch.2.11.0: 'SRR9990643' has 0 unresolved dependencies\n",
            "Read 10997899 spots for SRR9990643/SRR9990643.sra\n",
            "Written 10997899 spots for SRR9990643/SRR9990643.sra\n",
            "rm: cannot remove 'SRR9990643_3.fastq.gz': No such file or directory\n",
            "\n",
            "2021-04-12T22:43:35 prefetch.2.11.0: 1) Downloading 'SRR9990644'...\n",
            "2021-04-12T22:43:35 prefetch.2.11.0:  Downloading via HTTPS...\n",
            "2021-04-12T22:43:50 prefetch.2.11.0:  HTTPS download succeed\n",
            "2021-04-12T22:43:51 prefetch.2.11.0:  'SRR9990644' is valid\n",
            "2021-04-12T22:43:51 prefetch.2.11.0: 1) 'SRR9990644' was downloaded successfully\n",
            "2021-04-12T22:43:51 prefetch.2.11.0: 'SRR9990644' has 0 unresolved dependencies\n",
            "Read 10554424 spots for SRR9990644/SRR9990644.sra\n",
            "Written 10554424 spots for SRR9990644/SRR9990644.sra\n",
            "rm: cannot remove 'SRR9990644_3.fastq.gz': No such file or directory\n",
            "\n",
            "2021-04-12T22:55:29 prefetch.2.11.0: 1) Downloading 'SRR9990645'...\n",
            "2021-04-12T22:55:29 prefetch.2.11.0:  Downloading via HTTPS...\n",
            "2021-04-12T22:55:43 prefetch.2.11.0:  HTTPS download succeed\n",
            "2021-04-12T22:55:44 prefetch.2.11.0:  'SRR9990645' is valid\n",
            "2021-04-12T22:55:44 prefetch.2.11.0: 1) 'SRR9990645' was downloaded successfully\n",
            "2021-04-12T22:55:44 prefetch.2.11.0: 'SRR9990645' has 0 unresolved dependencies\n",
            "Read 11355017 spots for SRR9990645/SRR9990645.sra\n",
            "Written 11355017 spots for SRR9990645/SRR9990645.sra\n",
            "rm: cannot remove 'SRR9990645_3.fastq.gz': No such file or directory\n",
            "\n",
            "2021-04-12T23:07:53 prefetch.2.11.0: 1) Downloading 'SRR9990646'...\n",
            "2021-04-12T23:07:53 prefetch.2.11.0:  Downloading via HTTPS...\n",
            "2021-04-12T23:08:08 prefetch.2.11.0:  HTTPS download succeed\n",
            "2021-04-12T23:08:09 prefetch.2.11.0:  'SRR9990646' is valid\n",
            "2021-04-12T23:08:09 prefetch.2.11.0: 1) 'SRR9990646' was downloaded successfully\n",
            "2021-04-12T23:08:09 prefetch.2.11.0: 'SRR9990646' has 0 unresolved dependencies\n",
            "Read 10950939 spots for SRR9990646/SRR9990646.sra\n",
            "Written 10950939 spots for SRR9990646/SRR9990646.sra\n",
            "rm: cannot remove 'SRR9990646_3.fastq.gz': No such file or directory\n",
            "\n",
            "2021-04-12T23:19:54 prefetch.2.11.0: 1) Downloading 'SRR9990647'...\n",
            "2021-04-12T23:19:54 prefetch.2.11.0:  Downloading via HTTPS...\n",
            "2021-04-12T23:20:36 prefetch.2.11.0:  HTTPS download succeed\n",
            "2021-04-12T23:20:40 prefetch.2.11.0:  'SRR9990647' is valid\n",
            "2021-04-12T23:20:40 prefetch.2.11.0: 1) 'SRR9990647' was downloaded successfully\n",
            "2021-04-12T23:20:40 prefetch.2.11.0: 'SRR9990647' has 0 unresolved dependencies\n",
            "Read 12884052 spots for SRR9990647/SRR9990647.sra\n",
            "Written 12884052 spots for SRR9990647/SRR9990647.sra\n",
            "rm: cannot remove 'SRR9990647_3.fastq.gz': No such file or directory\n",
            "\n",
            "2021-04-12T23:40:59 prefetch.2.11.0: 1) Downloading 'SRR9990648'...\n",
            "2021-04-12T23:40:59 prefetch.2.11.0:  Downloading via HTTPS...\n",
            "2021-04-12T23:41:28 prefetch.2.11.0:  HTTPS download succeed\n",
            "2021-04-12T23:41:31 prefetch.2.11.0:  'SRR9990648' is valid\n",
            "2021-04-12T23:41:31 prefetch.2.11.0: 1) 'SRR9990648' was downloaded successfully\n",
            "2021-04-12T23:41:31 prefetch.2.11.0: 'SRR9990648' has 0 unresolved dependencies\n",
            "Read 10216311 spots for SRR9990648/SRR9990648.sra\n",
            "Written 10216311 spots for SRR9990648/SRR9990648.sra\n",
            "rm: cannot remove 'SRR9990648_3.fastq.gz': No such file or directory\n",
            "\n",
            "2021-04-12T23:58:04 prefetch.2.11.0: 1) Downloading 'SRR9990649'...\n",
            "2021-04-12T23:58:04 prefetch.2.11.0:  Downloading via HTTPS...\n",
            "2021-04-12T23:58:41 prefetch.2.11.0:  HTTPS download succeed\n",
            "2021-04-12T23:58:45 prefetch.2.11.0:  'SRR9990649' is valid\n",
            "2021-04-12T23:58:45 prefetch.2.11.0: 1) 'SRR9990649' was downloaded successfully\n",
            "2021-04-12T23:58:45 prefetch.2.11.0: 'SRR9990649' has 0 unresolved dependencies\n",
            "Read 13271228 spots for SRR9990649/SRR9990649.sra\n",
            "Written 13271228 spots for SRR9990649/SRR9990649.sra\n",
            "rm: cannot remove 'SRR9990649_3.fastq.gz': No such file or directory\n",
            "\n",
            "2021-04-13T00:19:30 prefetch.2.11.0: 1) Downloading 'SRR9990650'...\n",
            "2021-04-13T00:19:30 prefetch.2.11.0:  Downloading via HTTPS...\n",
            "2021-04-13T00:20:08 prefetch.2.11.0:  HTTPS download succeed\n",
            "2021-04-13T00:20:11 prefetch.2.11.0:  'SRR9990650' is valid\n",
            "2021-04-13T00:20:11 prefetch.2.11.0: 1) 'SRR9990650' was downloaded successfully\n",
            "2021-04-13T00:20:11 prefetch.2.11.0: 'SRR9990650' has 0 unresolved dependencies\n",
            "Read 10549316 spots for SRR9990650/SRR9990650.sra\n",
            "Written 10549316 spots for SRR9990650/SRR9990650.sra\n",
            "rm: cannot remove 'SRR9990650_3.fastq.gz': No such file or directory\n"
          ],
          "name": "stdout"
        }
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "bmN2opl1qHSh"
      },
      "source": [
        "# Define fastq list to feed kb\n",
        "\n",
        "fastqs = []\n",
        "for sra in SRAs:\n",
        "  for read in SRAs_to_keep:\n",
        "    fastqs.append(sra + \"_\" + read + \".fastq.gz\")"
      ],
      "execution_count": 4,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "5tT2sVv2-XMD"
      },
      "source": [
        "## Metadata"
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "ggNFAQJCRh6P"
      },
      "source": [
        "# define the values for the analysis\n",
        "\n",
        "# accession id for the data\n",
        "id = \"GSE135922\"\n",
        "\n",
        "no_samples = 1\n",
        "\n",
        "fastqs_per_sample = [8] \n",
        "\n",
        "sample_id = samp_id\n",
        "\n",
        "database_id = [id] * no_samples\n",
        "\n",
        "tissue = [\"eye\"] * no_samples\n",
        "\n",
        "species = [\"human\"] * no_samples\n",
        "\n",
        "technology = [\"10xv3\"] * no_samples\n",
        "\n",
        "paper = [\"Sungnak et al 2020\"] * no_samples\n",
        "\n",
        "figure = [\"Fig 1\"] * no_samples\n",
        "\n",
        "\n",
        "# Set string variables for kb functions\n",
        "\n",
        "species_kb = species[0]\n",
        "\n",
        "technology_kb = technology[0]"
      ],
      "execution_count": 5,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "FPwL0-_-KSAw"
      },
      "source": [
        "# Imports and installs"
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "1HSlUGZEp3oP",
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "outputId": "dbe07331-0c6e-41c3-9a87-da442d4856b2"
      },
      "source": [
        "# install and import necessary software\n",
        "\n",
        "# Install kb and scanpy\n",
        "!pip -q install kb-python \n",
        "!pip -q install scanpy\n",
        "\n",
        "import re\n",
        "import os\n",
        "\n",
        "# Setup\n",
        "\n",
        "import anndata\n",
        "import pandas as pd\n",
        "import numpy as np\n",
        "import matplotlib.pyplot as plt\n",
        "import matplotlib as mpl\n",
        "import matplotlib.patches as mpatches\n",
        "import scanpy as sc\n",
        "from scipy import stats\n",
        "\n",
        "from collections import OrderedDict\n",
        "from sklearn.decomposition import TruncatedSVD\n",
        "from sklearn.manifold import TSNE\n",
        "from sklearn.preprocessing import scale\n",
        "\n",
        "from sklearn.cluster import KMeans\n",
        "from sklearn.preprocessing import normalize\n",
        "from sklearn.preprocessing import LabelEncoder\n",
        "from sklearn.neighbors import NeighborhoodComponentsAnalysis\n",
        "from matplotlib import cm\n",
        "from matplotlib.lines import Line2D\n",
        "\n",
        "def nd(arr):\n",
        "    return np.asarray(arr).reshape(-1)\n",
        "def yex(ax):\n",
        "    lims = [np.min([ax.get_xlim(), ax.get_ylim()]),\n",
        "            np.max([ax.get_xlim(), ax.get_ylim()])]\n",
        "\n",
        "    # now plot both limits against eachother\n",
        "    ax.plot(lims, lims, 'k-', alpha=0.75, zorder=0)\n",
        "    ax.set_aspect('equal')\n",
        "    ax.set_xlim(lims)\n",
        "    ax.set_ylim(lims)\n",
        "    return ax\n",
        "\n",
        "def trim_axs(axs, N):\n",
        "    \"\"\"little helper to massage the axs list to have correct length...\"\"\"\n",
        "    axs = axs.flat\n",
        "    for ax in axs[N:]:\n",
        "        ax.remove()\n",
        "    return axs[:N]\n",
        "\n",
        "import warnings\n",
        "warnings.filterwarnings('ignore')\n",
        "\n",
        "fsize=20\n",
        "\n",
        "plt.rcParams.update({'font.size': fsize})\n",
        "%config InlineBackend.figure_format = 'retina'"
      ],
      "execution_count": 6,
      "outputs": [
        {
          "output_type": "stream",
          "text": [
            "\u001b[K     |████████████████████████████████| 59.1MB 107kB/s \n",
            "\u001b[K     |████████████████████████████████| 51kB 5.3MB/s \n",
            "\u001b[K     |████████████████████████████████| 133kB 43.7MB/s \n",
            "\u001b[K     |████████████████████████████████| 10.3MB 25.2MB/s \n",
            "\u001b[K     |████████████████████████████████| 13.2MB 182kB/s \n",
            "\u001b[K     |████████████████████████████████| 112kB 49.5MB/s \n",
            "\u001b[K     |████████████████████████████████| 81kB 8.3MB/s \n",
            "\u001b[K     |████████████████████████████████| 51kB 4.5MB/s \n",
            "\u001b[K     |████████████████████████████████| 1.2MB 41.2MB/s \n",
            "\u001b[K     |████████████████████████████████| 71kB 7.5MB/s \n",
            "\u001b[?25h  Building wheel for loompy (setup.py) ... \u001b[?25l\u001b[?25hdone\n",
            "  Building wheel for numpy-groupies (setup.py) ... \u001b[?25l\u001b[?25hdone\n",
            "  Building wheel for umap-learn (setup.py) ... \u001b[?25l\u001b[?25hdone\n",
            "  Building wheel for sinfo (setup.py) ... \u001b[?25l\u001b[?25hdone\n",
            "  Building wheel for pynndescent (setup.py) ... \u001b[?25l\u001b[?25hdone\n"
          ],
          "name": "stdout"
        }
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "7V6FbosVrvP4"
      },
      "source": [
        "# Downloads: index"
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "PC9oETW1pE9t"
      },
      "source": [
        "# Download the corresponding Kallisto index to fastq folder\n",
        "!kb ref -d $species_kb -i index.idx -g t2g.txt -f1 transcriptome.fasta"
      ],
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "VInyuq1Dp7iz"
      },
      "source": [
        "# Process fastq files (modify kb command according to fastqs list)\n"
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "F6FD32x-shPe"
      },
      "source": [
        "fastqs"
      ],
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "iXPeztS_HhCo",
        "outputId": "0c60dca5-2361-4065-a943-aea258d5c0d3",
        "colab": {
          "base_uri": "https://localhost:8080/"
        }
      },
      "source": [
        "if no_samples == 1:\n",
        "\n",
        "  # Write the kb count command as a string with all fastqs of the list as an input\n",
        "  cmd = \"kb count --h5ad -i index.idx -g t2g.txt -x \" + technology_kb + \" -o tccoutput\" + sample_id[0] + \" \"\\\n",
        "  + \"--filter bustools --tcc -t 2 --overwrite \" + \"'\" +  \"' '\".join(fastqs) + \"'\"\n",
        "  \n",
        "  # Execute it\n",
        "  !$cmd\n",
        "\n",
        "# If more than one sample, iterate through fastqs accordingly\n",
        "else:\n",
        "\n",
        "  # Initializa counter for fastq files\n",
        "  j = 0\n",
        "\n",
        "  # Loop over samples for analysis\n",
        "  for i in range(no_samples):\n",
        "\n",
        "    fastqs_to_analyze = fastqs[j:j + fastqs_per_sample[i]]\n",
        "    # Write the kb count command as a string\n",
        "    cmd = \"kb count --h5ad -i ../index.idx -g ../t2g.txt -x \" + technology_kb + \" -o tccoutput\" + sample_id[i] + \" \\\n",
        "    --filter bustools --tcc -t 2 --overwrite \" + \"'\" +  \"' '\".join(fastqs_to_analyze) + \"'\"\n",
        "\n",
        "    # Execute it\n",
        "    !$cmd\n",
        "\n",
        "    # Update j to move to the next set of fastq\n",
        "    j = j + fastqs_per_sample[i]\n",
        "\n",
        "\n",
        "\n",
        "\n",
        "\n",
        "\n",
        "results = {}\n",
        "tcc_results = {}\n",
        "for sample in sample_id:\n",
        "  # Read the filtered gene count matrix\n",
        "  results[\"data_\" + sample] = anndata.read(\"/content/result\" + sample)\n",
        "\n",
        "\n",
        "for sample in sample_id:\n",
        "  output = \"bus_raw_\" + sample + \"_from_gc\"\n",
        "  !mkdir $output\n",
        "  folder_tcc = \"tccoutput\" + sample\n",
        "\n",
        "  # read tcc matrix\n",
        "  tcc_results[sample] = anndata.read(folder_tcc + \"/counts_unfiltered/adata.h5ad\")\n",
        "  # apply gene count matrix's filter\n",
        "  tcc_results[sample] = tcc_results[sample][results[\"data_\" + sample].obs.index.values]\n",
        "  # transfer obs data\n",
        "  tcc_results[sample].obs = results[\"data_\" + sample].obs\n",
        "  # transfer metadata\n",
        "  tcc_results[sample].uns = results[\"data_\" + sample].uns\n",
        "  # write tcc matrix\n",
        "  tcc_results[sample].write(\"/content/tcc_\" + sample)\n",
        "  #gzip tcc matrix\n",
        "  cmd = \"gzip /content/tcc_\" + sample\n",
        "  !$cmd\n",
        "  #gzip unfiltered bus file\n",
        "  cmd = \"gzip \" + folder_tcc + \"/output.unfiltered.bus\"\n",
        "  !$cmd\n",
        "  # move and re-name bus file with sample id\n",
        "  cmd = \"mv \" + folder_tcc + \"/output.unfiltered.bus.gz \" + output + \"/\" + sample +\".unfiltered.bus.gz\"\n",
        "  !$cmd\n",
        "  # gzip all unfiltered counts\n",
        "  cmd = \"gzip \" + folder_tcc + \"/counts_unfiltered/*\"\n",
        "  !$cmd\n",
        "\n",
        "  # move unfiltered counts for both\n",
        "  cmd = \"mv \" + folder_tcc + \"/counts_unfiltered \" + output + \"/tcc_unfiltered\"\n",
        "  !$cmd\n",
        "\n",
        "  cmd = \"mv \" + folder_tcc + \"/*.json \" + output \n",
        "  !$cmd\n",
        "  # zip all files\n",
        "  cmd = \"zip -r \"+ output + \".zip \" + output\n",
        "  !$cmd\n",
        "\n",
        "\n",
        "import time\n",
        "time.sleep(60000000)"
      ],
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "text": [
            "[2021-04-13 02:45:00,945]    INFO Using index index.idx to generate BUS file to tccoutputGSM4037981 from\n",
            "[2021-04-13 02:45:00,945]    INFO         SRR9990627_1.fastq.gz\n",
            "[2021-04-13 02:45:00,945]    INFO         SRR9990627_2.fastq.gz\n",
            "[2021-04-13 02:45:00,945]    INFO         SRR9990628_1.fastq.gz\n",
            "[2021-04-13 02:45:00,945]    INFO         SRR9990628_2.fastq.gz\n",
            "[2021-04-13 02:45:00,945]    INFO         SRR9990629_1.fastq.gz\n",
            "[2021-04-13 02:45:00,945]    INFO         SRR9990629_2.fastq.gz\n",
            "[2021-04-13 02:45:00,945]    INFO         SRR9990630_1.fastq.gz\n",
            "[2021-04-13 02:45:00,945]    INFO         SRR9990630_2.fastq.gz\n",
            "[2021-04-13 02:45:00,945]    INFO         SRR9990631_1.fastq.gz\n",
            "[2021-04-13 02:45:00,945]    INFO         SRR9990631_2.fastq.gz\n",
            "[2021-04-13 02:45:00,945]    INFO         SRR9990632_1.fastq.gz\n",
            "[2021-04-13 02:45:00,945]    INFO         SRR9990632_2.fastq.gz\n",
            "[2021-04-13 02:45:00,945]    INFO         SRR9990633_1.fastq.gz\n",
            "[2021-04-13 02:45:00,946]    INFO         SRR9990633_2.fastq.gz\n",
            "[2021-04-13 02:45:00,946]    INFO         SRR9990634_1.fastq.gz\n",
            "[2021-04-13 02:45:00,946]    INFO         SRR9990634_2.fastq.gz\n",
            "[2021-04-13 02:45:00,946]    INFO         SRR9990635_1.fastq.gz\n",
            "[2021-04-13 02:45:00,946]    INFO         SRR9990635_2.fastq.gz\n",
            "[2021-04-13 02:45:00,946]    INFO         SRR9990636_1.fastq.gz\n",
            "[2021-04-13 02:45:00,946]    INFO         SRR9990636_2.fastq.gz\n",
            "[2021-04-13 02:45:00,946]    INFO         SRR9990637_1.fastq.gz\n",
            "[2021-04-13 02:45:00,946]    INFO         SRR9990637_2.fastq.gz\n",
            "[2021-04-13 02:45:00,946]    INFO         SRR9990638_1.fastq.gz\n",
            "[2021-04-13 02:45:00,946]    INFO         SRR9990638_2.fastq.gz\n",
            "[2021-04-13 02:45:00,946]    INFO         SRR9990639_1.fastq.gz\n",
            "[2021-04-13 02:45:00,946]    INFO         SRR9990639_2.fastq.gz\n",
            "[2021-04-13 02:45:00,946]    INFO         SRR9990640_1.fastq.gz\n",
            "[2021-04-13 02:45:00,946]    INFO         SRR9990640_2.fastq.gz\n",
            "[2021-04-13 02:45:00,946]    INFO         SRR9990641_1.fastq.gz\n",
            "[2021-04-13 02:45:00,946]    INFO         SRR9990641_2.fastq.gz\n",
            "[2021-04-13 02:45:00,946]    INFO         SRR9990642_1.fastq.gz\n",
            "[2021-04-13 02:45:00,947]    INFO         SRR9990642_2.fastq.gz\n",
            "[2021-04-13 02:45:00,947]    INFO         SRR9990643_1.fastq.gz\n",
            "[2021-04-13 02:45:00,947]    INFO         SRR9990643_2.fastq.gz\n",
            "[2021-04-13 02:45:00,947]    INFO         SRR9990644_1.fastq.gz\n",
            "[2021-04-13 02:45:00,947]    INFO         SRR9990644_2.fastq.gz\n",
            "[2021-04-13 02:45:00,947]    INFO         SRR9990645_1.fastq.gz\n",
            "[2021-04-13 02:45:00,947]    INFO         SRR9990645_2.fastq.gz\n",
            "[2021-04-13 02:45:00,947]    INFO         SRR9990646_1.fastq.gz\n",
            "[2021-04-13 02:45:00,947]    INFO         SRR9990646_2.fastq.gz\n",
            "[2021-04-13 02:45:00,947]    INFO         SRR9990647_1.fastq.gz\n",
            "[2021-04-13 02:45:00,947]    INFO         SRR9990647_2.fastq.gz\n",
            "[2021-04-13 02:45:00,947]    INFO         SRR9990648_1.fastq.gz\n",
            "[2021-04-13 02:45:00,947]    INFO         SRR9990648_2.fastq.gz\n",
            "[2021-04-13 02:45:00,947]    INFO         SRR9990649_1.fastq.gz\n",
            "[2021-04-13 02:45:00,947]    INFO         SRR9990649_2.fastq.gz\n",
            "[2021-04-13 02:45:00,947]    INFO         SRR9990650_1.fastq.gz\n",
            "[2021-04-13 02:45:00,947]    INFO         SRR9990650_2.fastq.gz\n",
            "[2021-04-13 03:27:48,521]    INFO Sorting BUS file tccoutputGSM4037981/output.bus to tccoutputGSM4037981/tmp/output.s.bus\n",
            "[2021-04-13 03:28:26,784]    INFO Whitelist not provided\n",
            "[2021-04-13 03:28:26,784]    INFO Copying pre-packaged 10XV3 whitelist to tccoutputGSM4037981\n",
            "[2021-04-13 03:28:27,758]    INFO Inspecting BUS file tccoutputGSM4037981/tmp/output.s.bus\n",
            "[2021-04-13 03:28:44,566]    INFO Correcting BUS records in tccoutputGSM4037981/tmp/output.s.bus to tccoutputGSM4037981/tmp/output.s.c.bus with whitelist tccoutputGSM4037981/10xv3_whitelist.txt\n",
            "[2021-04-13 03:29:03,592]    INFO Sorting BUS file tccoutputGSM4037981/tmp/output.s.c.bus to tccoutputGSM4037981/output.unfiltered.bus\n",
            "[2021-04-13 03:29:21,593]    INFO Generating count matrix tccoutputGSM4037981/counts_unfiltered/cells_x_tcc from BUS file tccoutputGSM4037981/output.unfiltered.bus\n",
            "[2021-04-13 03:29:43,918]    INFO Reading matrix tccoutputGSM4037981/counts_unfiltered/cells_x_tcc.mtx\n",
            "[2021-04-13 03:30:07,937]    INFO Writing matrix to h5ad tccoutputGSM4037981/counts_unfiltered/adata.h5ad\n",
            "[2021-04-13 03:30:14,684]    INFO Filtering with bustools\n",
            "[2021-04-13 03:30:14,685]    INFO Generating whitelist tccoutputGSM4037981/filter_barcodes.txt from BUS file tccoutputGSM4037981/output.unfiltered.bus\n",
            "[2021-04-13 03:30:14,907]    INFO Correcting BUS records in tccoutputGSM4037981/output.unfiltered.bus to tccoutputGSM4037981/tmp/output.unfiltered.c.bus with whitelist tccoutputGSM4037981/filter_barcodes.txt\n",
            "[2021-04-13 03:30:24,885]    INFO Sorting BUS file tccoutputGSM4037981/tmp/output.unfiltered.c.bus to tccoutputGSM4037981/output.filtered.bus\n",
            "[2021-04-13 03:30:43,900]    INFO Generating count matrix tccoutputGSM4037981/counts_filtered/cells_x_tcc from BUS file tccoutputGSM4037981/output.filtered.bus\n",
            "[2021-04-13 03:31:02,513]    INFO Reading matrix tccoutputGSM4037981/counts_filtered/cells_x_tcc.mtx\n",
            "[2021-04-13 03:31:24,696]    INFO Writing matrix to h5ad tccoutputGSM4037981/counts_filtered/adata.h5ad\n"
          ],
          "name": "stdout"
        },
        {
          "output_type": "stream",
          "text": [
            "Variable names are not unique. To make them unique, call `.var_names_make_unique`.\n"
          ],
          "name": "stderr"
        },
        {
          "output_type": "stream",
          "text": [
            "  adding: bus_raw_GSM4037981_from_gc/ (stored 0%)\n",
            "  adding: bus_raw_GSM4037981_from_gc/tcc_unfiltered/ (stored 0%)\n",
            "  adding: bus_raw_GSM4037981_from_gc/tcc_unfiltered/adata.h5ad.gz (deflated 8%)\n",
            "  adding: bus_raw_GSM4037981_from_gc/tcc_unfiltered/cells_x_tcc.mtx.gz (deflated 0%)\n",
            "  adding: bus_raw_GSM4037981_from_gc/tcc_unfiltered/cells_x_tcc.barcodes.txt.gz (deflated 1%)\n",
            "  adding: bus_raw_GSM4037981_from_gc/tcc_unfiltered/cells_x_tcc.ec.txt.gz (deflated 2%)\n",
            "  adding: bus_raw_GSM4037981_from_gc/inspect.json (deflated 57%)\n",
            "  adding: bus_raw_GSM4037981_from_gc/GSM4037981.unfiltered.bus.gz (deflated 3%)\n",
            "  adding: bus_raw_GSM4037981_from_gc/run_info.json (deflated 73%)\n",
            "  adding: bus_raw_GSM4037981_from_gc/kb_info.json (deflated 82%)\n"
          ],
          "name": "stdout"
        }
      ]
    }
  ]
}